File: analysis/api/utils/clustering.py
#=================================================================================================
# Project: CADS/MADS - An Integrated Web-based Visual Platform for Materials Informatics
#          Hokkaido University (2018)
#          Last Update: Q3 2023
# ________________________________________________________________________________________________
# Authors: Mikael Nicander Kuwahara (Lead Developer) [2021-]
#          Jun Fujima (Former Lead Developer) [2018-2021]
# ________________________________________________________________________________________________
# Description: Serverside (Django) rest api utils for the 'Analysis' page involving
#              'clustering' components
# ------------------------------------------------------------------------------------------------
# Notes:  This is one of the REST API parts of the 'analysis' interface of the website that
#         allows serverside work for the 'clustering' component.
# ------------------------------------------------------------------------------------------------
# References: logging, numpy, pandas and sklearn libs
#=================================================================================================

#-------------------------------------------------------------------------------------------------
# Import required Libraries
#-------------------------------------------------------------------------------------------------
import logging
import numpy as np
import pandas as pd
from sklearn.cluster import KMeans
from sklearn.mixture import GaussianMixture

logger = logging.getLogger(__name__)
#-------------------------------------------------------------------------------------------------


#-------------------------------------------------------------------------------------------------
def get_clusters(data):
    vis_type = data['view']['settings']['visType']

    result = {}
    result['vis_type'] = vis_type

    feature_columns = data['view']['settings']['featureColumns']
    num_of_clusters = int(data['view']['settings']['numberOfClusters'])
    if num_of_clusters > 10:
        num_of_clusters = 10

    dataset = data['data']
    df = pd.DataFrame(dataset)
    df_target = df[feature_columns]
    X = df_target.values

    if(vis_type == "Bar Chart"):
        method = data['view']['settings']['method']

        cif = None
        if (method == 'KMeans'):
            clf = KMeans(n_clusters=num_of_clusters)
        else:
            clf = GaussianMixture(n_components=num_of_clusters)

        clf.fit(X)
        y = clf.predict(X)

        result['cluster'] = y
    else:
        kmean = KMeans(n_clusters=num_of_clusters)
        kmean.fit(X)
        y = kmean.predict(X)

        result['cluster'] = y
        result['data'] = data['data']


    return result
#-------------------------------------------------------------------------------------------------

File: assets/js/app/mads-cmv/components/Views/Clustering.js
/*=================================================================================================
// Project: CADS/MADS - An Integrated Web-based Visual Platform for Materials Informatics
//          Hokkaido University (2018)
//          Last Update: Q3 2023
// ________________________________________________________________________________________________
// Authors: Mikael Nicander Kuwahara (Lead Developer) [2021-]
//          Jun Fujima (Former Lead Developer) [2018-2021]
// ________________________________________________________________________________________________
// Description: This is the Inner workings and Content Manager Controler of the
//              'K-Means Clustering' View
// ------------------------------------------------------------------------------------------------
// Notes: 'K-Means Clustering' is the manager of all current input that controls the final view of the
//         this specific case of a 'BarChart' visualization component.
// ------------------------------------------------------------------------------------------------
// References: 3rd party pandas libs, Internal ViewWrapper & Form Utility Support,
//             Internal PieChart & PieForm libs,
=================================================================================================*/


//-------------------------------------------------------------------------------------------------
// Load required libraries
//-------------------------------------------------------------------------------------------------
import { DataFrame } from 'pandas-js';

import withCommandInterface from './ViewWrapper';
import convertExtentValues from './FormUtils';

import ClusteringVis from '../VisComponents/ClusteringVis';
import ClusteringForm from './ClusteringForm';

//-------------------------------------------------------------------------------------------------


//-------------------------------------------------------------------------------------------------
// The View Class for this Visualization Component
//-------------------------------------------------------------------------------------------------
export default class ClusteringView extends withCommandInterface( ClusteringVis, ClusteringForm ) {

  // Manages data selection changes in the view
  handleSelectionChange = (indices) => {
    if (indices.length === 0) {
      return;
    }

    const { dataset, updateSelection, id } = this.props;

    let selections = [];
    const { cluster } = dataset[id];
    indices.forEach((cid) => {
      const sel = cluster.forEach((i, ii) => {
        if (i === cid) {
          selections.push(ii);
        }
      });
    });

    updateSelection(selections);
  };

  // Manages config settings changes (passed by the connected form) in the view
  handleSubmit = (values) => {
    const { id, view, colorTags, actions, dataset } = this.props;
    let newValues = { ...values };

    // filter out non-existing columns & colorTags
    if (values.filter) {
      const colorTagIds = colorTags.map((c) => c.id);
      const filteredFilters = values.filter.filter((f) =>
        colorTagIds.includes(f)
      );
      newValues.filter = filteredFilters;
    }

    var data = {};
    if(newValues.visType == "Bar Chart"){
      // filter out featureColumns
      const columns = this.getColumnOptionArray();
      if (values.featureColumns) {
        const filteredColumns = values.featureColumns.filter((f) =>
          columns.includes(f)
        );
        newValues.featureColumns = filteredColumns;
      }

      // extract data
      const df = new DataFrame(dataset.main.data);
      newValues.featureColumns.forEach((c) => {
        const fc = df.get(c);
        data[c] = fc.values.toArray();
      });
    }
    else{
      if (!values.colorAssignmentEnabled) { newValues.mappings.color = ''; }
      data = [...dataset.main.data];
    }

    newValues = convertExtentValues(newValues);

    actions.sendRequestViewUpdate(view, newValues, data);
  };

  // Manages data changes in the view
  mapData = (dataset) => {
    const { id, view } = this.props;
    const data = {};
    const counts = [];

    if (dataset[id]) {
      var testValue = this.props.view.settings.featureColumns[0]
      if (dataset.main.schema.fields.some(e => e.name === testValue)) {
        data.visType = dataset[id].vis_type

        if(data.visType == "Bar Chart"){
          const cids = dataset[id].cluster;
          const cidsStr = [];
          const noc = view.settings.numberOfClusters;
          for (let i = 0; i < noc; i++) {
            const c = cids.filter((x) => x === i).length;
            counts.push(c);
            cidsStr.push(i.toString());
          }

          data.cids = cidsStr;
          data.counts = counts;
        }
        else{
          data.cluster = dataset[id].cluster;
          data.data = dataset[id].data;
        }
      }
      else{
        data["resetRequest"] = true;
      }


    }

    return data;
  };
}
//-------------------------------------------------------------------------------------------------
